### PR TITLE
2D Matrix Rotation

### PR DESCRIPTION
**Pull request by vanshikachokhani**
#### Changes made
I have written an logic to rotate a matrix by 90 degrees in clockwise direction in C++
This is a small contribution as a part of GSSoC21.

I have tested the code and it works fine for every test case and included some sample test cases and also the complete explanation of the algorithm.

### CODE REVIEW

File: Code/C++/2d_matrix_rotation_90degree_clockwise.cpp [hunk 1]
```diff
@@ -0,0 +1,123 @@
...
+3 4 2 1 
+7 6 5 4 
+3 5 6 7 
+
+Matrix rotated by 90degrees
+3 7 3 5 
+5 6 4 8 
+6 5 2 9 
+7 4 1 7 
+
```
Comment (rudrakshi99) [inline Code/C++/2d_matrix_rotation_90degree_clockwise.cpp:122]: add time and space complexities

**Review by rudrakshi99 — CHANGES_REQUESTED** [review]

**Comment by rudrakshi99** [comment]
write all the comments just above the code

**Comment by vanshikachokhani** [comment]
> write all the comments just above the code

Yess I will do the required changes and update it.

**Review by raksha009 — COMMENTED** [review]
Link our code with the appropriate folder,

**Comment by vanshikachokhani** [comment]
> Link our code with the appropriate folder,

I have done the requested changes. Please review it.

File: Code/C++/2d_matrix_rotation_90degree_clockwise.cpp (hunk 1 above)
Comment (vanshikachokhani) [reply Code/C++/2d_matrix_rotation_90degree_clockwise.cpp:144]: > add time and space complexities

I have added the time and space complexities.

File: Code/C++/2d_matrix_rotation_90degree_clockwise.cpp [hunk 2]
```diff
@@ -19,45 +19,56 @@ Example: The given matrix is
             7 4 1
             8 5 2
             9 6 3
+        
+Time Complexity: O(m*n) where m and n are no of rows and columns respectively
+Space Complexity: O(1)
```
Comment (rudrakshi99) [inline Code/C++/2d_matrix_rotation_90degree_clockwise.cpp:24]: add complexities at the end of the code.

**Review by rudrakshi99 — CHANGES_REQUESTED** [review]

**Comment by raksha009** [comment]
> > Link our code with the appropriate folder,
> 
> I have done the requested changes. Please review it.

Heyy, you need to update the required readme to update the code. Plz go through the guidelines and the pinned issues : )

**Comment by vanshikachokhani** [comment]
> > > Link our code with the appropriate folder,
> > 
> > 
> > I have done the requested changes. Please review it.
> 
> Heyy, you need to update the required readme to update the code. Plz go through the guidelines and the pinned issues : )

I have updated the readme. Please merge the branch if no further changes required.

**Comment by vanshikachokhani** [comment]
> > Link our code with the appropriate folder,
> 
> I have done the requested changes. Please review it.

I have done the requested changes

File: Code/C++/2d_matrix_rotation_90degree_clockwise.cpp [hunk 3]
```diff
@@ -0,0 +1,148 @@
+/*
```
Comment (raksha009) [inline Code/C++/2d_matrix_rotation_90degree_clockwise.cpp:1]: Include the entire description in a single multiline comment.

File: Code/C++/2d_matrix_rotation_90degree_clockwise.cpp [hunk 4]
```diff
@@ -0,0 +1,148 @@
...
+*/ 
+        
+#include<bits/stdc++.h>
+using namespace std;
+#define int long long
+
+//utility function to carry out the transpose of a 2d matrix
+void transpose(vector<vector<int>>& v)                   
+{
+    int n = v.size();
```
Comment (raksha009) [inline Code/C++/2d_matrix_rotation_90degree_clockwise.cpp:32]: Plz specify in comments what is n?

File: Code/C++/2d_matrix_rotation_90degree_clockwise.cpp [hunk 5]
```diff
@@ -0,0 +1,148 @@
...
+}
+
+//utility function to take reflection of a 2d matrix along the column
+void reflect(vector<vector<int>>& v)                    
+{
+    int n = v.size();
+
+    for(int i = 0 ; i < n; i++)
+    {
+        for(int j=0;j<v[i].size()/2;j++)
```
Comment (raksha009) [inline Code/C++/2d_matrix_rotation_90degree_clockwise.cpp:51]:
```suggestion
        for(int j = 0; j < v[i].size() / 2; j++)
```

Add space on both sides of operators like <,>,= etc

File: Code/C++/2d_matrix_rotation_90degree_clockwise.cpp [hunk 6]
```diff
@@ -0,0 +1,148 @@
...
+    int n;
+
+    //input size of matrix
+    cin>>n;                         
+
+    //declaring a 2d vector and assigning n number of rows
+    vector<vector<int>> v(n);       
+
+    //input elemets of matrix
+    for(int i=0;i<n;i++)
```
Comment (raksha009) [inline Code/C++/2d_matrix_rotation_90degree_clockwise.cpp:95]:
```suggestion
    for(int i=0;i<n;i++) {           
```
Start brackets like  this throughout!

File: Code/C++/2d_matrix_rotation_90degree_clockwise.cpp [hunk 7]
```diff
@@ -0,0 +1,148 @@
...
+    //print function call to print rotated matrix
+    print(v);                       
+
+    return 0;
+    
+}
+
+/* 
+
+Time Complexity: O(m*n) where m and n are no of rows and columns respectively
```
Comment (raksha009) [inline Code/C++/2d_matrix_rotation_90degree_clockwise.cpp:124]: Add time complexity after test cases

**Review by raksha009 — CHANGES_REQUESTED** [review]

File: Code/C++/2d_matrix_rotation_90degree_clockwise.cpp [hunk 8]
```diff
@@ -121,8 +106,7 @@ signed main()
 
 /* 
 
-Time Complexity: O(m*n) where m and n are no of rows and columns respectively
-Space Complexity: O(1)
+
```
Comment (rudrakshi99) [inline Code/C++/2d_matrix_rotation_90degree_clockwise.cpp:109]: use the input-output format in comments for testcase

**Review by rudrakshi99 — CHANGES_REQUESTED** [review]

**Review by raksha009 — APPROVED** [review]

**Pull request merged**